backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/InteropCallConvertors.kt

```diff
@@ -154,40 +162,165 @@ private fun InteropCallContext.convertIntegralToEnum(value: IrExpression, enumTy
+        accessor.isSetter -> writeEnumValueToMemory(nativePtr, callSite.getValueArgument(0)!!)
+        else -> error("")
+    }
+}
+
+private fun InteropCallContext.generateMemberAtAccess(callSite: IrCall): IrExpression {
+    val accessor = callSite.symbol.owner
+    val memberAt = accessor.getAnnotation(RuntimeNames.cStructMemberAt)!!
+    val offset = (memberAt.getValueArgument(0) as IrConst<Long>).value
+    val fieldPointer = calculateFieldPointer(callSite.dispatchReceiver!!, offset)
     return when {
         accessor.isGetter -> {
-            val enumClass = accessor.returnType.getClass()!!
-            val enumPrimitiveType = enumClass.properties.single { it.name.asString() == "value" }
-                    .getter!!.returnType
-            val readMemory = readPrimitiveFromMemory(nativePtr, enumPrimitiveType)
-            return convertIntegralToEnum(readMemory, accessor.returnType)
+            val type = accessor.returnType
+            when {
+                type.isCEnumType() -> readEnumValueFromMemory(fieldPointer, type)
+                type.isPrimitiveType() -> readPrimitiveFromMemory(fieldPointer, type)
+                type.isCPointer() -> readPointerFromMemory(fieldPointer)
+                type.isNativePointed() -> readPointedFromMemory(fieldPointer)
+                else -> error("Cannot get field type: ${type.getClass()?.name}")
+            }
         }
         accessor.isSetter -> {
-            val arg = callSite.getValueArgument(0)!!
-            val valueToWrite = convertEnumToIntegral(arg)
-            writePrimitiveToMemory(nativePtr, valueToWrite)
+            val value = callSite.getValueArgument(0)!!
+            when {
+                value.type.isCEnumType() -> writeEnumValueToMemory(fieldPointer, value)
+                value.type.isPrimitiveType() -> writePrimitiveToMemory(fieldPointer, value)
+                else -> error("Cannot set field of type ${value.type.getClass()?.name}")
+            }
         }
         else -> error("")
     }
+}
+
+private fun InteropCallContext.writeBits(
+        base: IrExpression,
+        offset: Long,
+        size: Int,
+        value: IrExpression
+): IrExpression {
+    val integralValue = when {
+        value.type.isCEnumType() -> convertEnumToIntegral(value)
+        else -> value
+    }
+    val targetType = symbols.writeBits.owner.valueParameters.last().type
+    val valueToWrite = castPrimitiveIfNeeded(integralValue, targetType)
+    return with(builder) {
+        irCall(symbols.writeBits).also {
+            it.putValueArgument(0, base)
+            it.putValueArgument(1, irLong(offset))
+            it.putValueArgument(2, irInt(size))
+            it.putValueArgument(3, valueToWrite)
+        }
+    }
+}
+
+private fun InteropCallContext.readBits(
+        base: IrExpression,
+        offset: Long,
+        size: Int,
+        type: IrType
+): IrExpression {
+    val isSigned = !type.isUnsigned()
```

Comment:
What if `type` is enum?

Reply:
Btw, do we have a test that would discover that bug?

Reply:
`interop_bitfields` covers this case, so yes. Looks like I forgot to turn `-mode metadata` on for this particular test.

Reply:
It seems to contain only one enum. We probably should have both signed and unsigned enum bitfield tests.